feat(compiler): add support for compile-time required inputs

crisbeto opened the pull request:
Adds support for marking a directive input as required. During template type checking, the compiler will verify that all required inputs have been specified and will raise a diagnostic if one or more are missing. Some specifics:
* Inputs are marked as required by passing an object literal with a `required: true` property to the `Input` decorator or into the `inputs` array.
* Required inputs imply that the directive can't work without them. This is why there's a new check that enforces that all required inputs of a host directive are exposed on the host.
* Required input diagnostics are reported through the `OutOfBandDiagnosticRecorder`, rather than generating a new structure in the TCB, because it allows us to provide a better error message.
* Currently required inputs are only supported during AOT compilation, because knowing which bindings are present during JIT can be tricky and may lead to increased bundle sizes.

Fixes #37706.

crisbeto commented on goldens/public-api/core/index.md:
```diff
@@ -793,17 +797,24 @@ export interface InjectorType<T> extends Type<T> {
 
 // @public
 export interface Input {
-    bindingPropertyName?: string;
```
I always found the `bindingPropertyName` terminology kind of confusing so I took the chance to rename it to `publicName` which (IMO) is easier to understand. It's also a bit shorter.

splincode replied:
why not just `alias`, or `aliasName`, `aliasAttr`?)

Harpush commented:
How will it affect `strictNullChecks` flag? I believe we still need the `!` operator? So instead of runtime checks in `ngOnInit` we can have the compiler throw errors for us. Anyway that's great! Thanks!

crisbeto commented:
Note for reviewer: the `test_aio_local` failure was triggered by these changes, but it's a pre-existing issue. It'll be resolved by #49293.

alxhub commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -91,6 +91,11 @@ export interface R3DirectiveMetadata {
+  /**
+   * Names of the class fields that are marked as required inputs.
+   */
+  requiredInputs: Set<string>|null;
```
We should blend this into the `inputs` field itself (maybe declaring an interface `R3InputMetadata` so `inputs` becomes `Record<string, R3InputMetadata>`. That way we're not having to pass around extra information about inputs separately from the rest of the metadata.

It would be good to do this as a refactoring commit prior to the required inputs implementation, so all that commit needs to do is add `required: boolean` to the `R3InputMetadata` type that's already in place, and we don't mix the cleanup changes in with the main implementation.

crisbeto replied:
That's what I wanted to do initially, but the problem is that we're using a special data structure for inputs and outputs throughout the codebase. I was considering changing the data structure to add a required field, but it felt like I was making it do too much.

alxhub commented on packages/core/src/metadata/directives.ts:
```diff
@@ -787,8 +793,8 @@ export interface InputDecorator {
    */
-  (bindingPropertyName?: string): any;
-  new(bindingPropertyName?: string): any;
+  (arg?: string|{publicName?: string, required?: boolean}): any;
```
We could use the `Input` interface itself as the type here:
```suggestion
  (arg?: string|Input): any;
```

jelbourn commented on packages/compiler-cli/src/ngtsc/annotations/common/src/diagnostics.ts:
```diff
@@ -227,6 +238,26 @@ function validateHostDirectiveMappings(
     }
   }
+
+  if (requiredBindings !== null && requiredBindings.size !== exposedRequiredBindings.size) {
```
Will there be a problem if you have a template like this?

```html
<x-y [thing]="'abc'" [thing]="'xyz'"></x-y>
```

crisbeto replied:
I think that this is handled during template parsing, I would have to double check.

jelbourn commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -18,13 +18,18 @@ import {HandlerFlags} from '../../../transform';
   'ViewChildren',
 ]);
 
+type StringMap<T> = {
+  [key: string]: T;
+};
```
Could you use the `Record` built-in? At the very least `Record<string, T>`

crisbeto commented on goldens/public-api/core/index.md:
```diff
@@ -456,7 +456,11 @@ export interface Directive {
-    inputs?: string[];
+    inputs?: (string | {
+        name: string;
+        publicName?: string;
```
I was calling it alias while working on the prototype, but it felt too vague. I don't mind changing back to it though.

elgreco247 commented:
> * Required inputs imply that the directive can't work without them. This is why there's a new check that enforces that all required inputs of a host directive are exposed on the host.

As described by @nartc in this tweet (https://twitter.com/Nartc1410/status/1631694281298640896?s=20) this enforcement might be a little bit too rigid concerning to the reusability of a directive.

Maybe the configuration object of host directives can be augmented with a property specifying whether the exposure of required inputs should be enforced.

For example `{directive: HostDir, inputs: ['inputAlias'], enforceRequiredInputs: false}`.

kbrilla commented on packages/compiler-cli/src/ngtsc/typecheck/test/diagnostics_spec.ts:
```diff
@@ -856,6 +856,220 @@ class TestComponent {
+      expect(messages).toEqual([]);
+    });
+
+    it('should not produce a diagnostic for a required input that is the same the directive selector',
```
Why no error in such a case? It will discourage from using single input directives in such a case as for example `<span icon></span>` here `icon` input should be required as this directive does not make sense without specifying what icon to display but as I understand it won't show diagnostic even if we mark it as required.  What more people will probably use some selector hacking/engenering to still get diagnostic - as for example 
[stackblitz](https://stackblitz.com/edit/angular-zmjfrf?file=src/main.ts)

kbrilla replied:
If that's the case at least some diagnostic message that required will not be applied would be useful, Still, that is not a small downside

kbrilla replied:
ok, so with `strictTemplates ` the only problem will be when input is of type `string`  because in other cases, for example `number ` there will be different compile time error: `Type 'string' is not assignable to type 'number'.`

kbrilla commented on packages/compiler-cli/test/ngtsc/host_directives_spec.ts:
```diff
@@ -777,6 +840,110 @@ runInEachFileSystem(() => {
            expect(messages).toEqual([]);
          });
+
+      it('should produce a diagnostic if a required input is not exposed on the host', () => {
```
I don't think that should be the case, as we can have a host directive lets again say [icon] that have required input which we will set in programmatically for example in constructor. For example 

```
@Directive({
  selector: '[icon]',
  standalone: true,
})
export class IconDirective {
  @Input({ required: true }) icon: string;
}

@Directive({
  selector: '[closeIcon]',
  hostDirectives: [IconDirective],
})
export class CloseIconDirective {
  iconDirective = inject(IconDirective);

  constructor() {
    this.iconDirective.icon = 'close';
  }
}
```

crisbeto replied:
The thinking with required inputs is that the directive *cannot* work without them. This applies to host directives as well.

elgreco247 replied:
@kbrilla 
This could be a solution:

```javascript
// Use this one in host directives
@Directive({
  standalone: true,
})
export class IconDirective {
  @Input() iconName: string;
}

// Use this one in templates
@Directive({
  selector: '[icon]',
  standalone: true,
  inputs: [{name: 'iconName', required: true}
})
export class IconDirectiveForTemplates extends IconDirective {}
```

kbrilla replied:
@elgreco247 but then for example libraries would have to export two directives one for use in host directives and another in templates. That would be a real mess.

crisbeto commented:
I've integrated the changes from #49333 and addressed the remaining feedback.

crisbeto committed:
feat(compiler): add support for compile-time required inputs

Adds support for marking a directive input as required. During template type checking, the compiler will verify that all required inputs have been specified and will raise a diagnostic if one or more are missing. Some specifics:
* Inputs are marked as required by passing an object literal with a `required: true` property to the `Input` decorator or into the `inputs` array.
* Required inputs imply that the directive can't work without them. This is why there's a new check that enforces that all required inputs of a host directive are exposed on the host.
* Required input diagnostics are reported through the `OutOfBandDiagnosticRecorder`, rather than generating a new structure in the TCB, because it allows us to provide a better error message.
* Currently required inputs are only supported during AOT compilation, because knowing which bindings are present during JIT can be tricky and may lead to increased bundle sizes.

Fixes #37706.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

alxhub reviewed (approved):
Reviewed-for: public-api

jessicajaniuk reviewed (approved):
reviewed-for: public-api

crisbeto commented:
Carertaker note: one target is failing internally. It doesn't look related to this change.

alxhub commented:
This PR was merged into the repository by commit 1a6ca68154dd73bac4b8d2e094d97952f60b3e30.

e-oz commented:
Are you sure it's merged into 16.0.0-next.3? 
I can't find it in 16.0.0-next.3. When I use @Input({required: true}) in a directive or a component:
```
error TS2345: Argument of type '{ required: boolean; }' is not assignable to parameter of type 'string'.
```

kbrilla commented:
it was reverted for some internal issues

pt., 17 mar 2023 o 12:15 Evgeniy OZ ***@***.***> napisał(a):

> Are you sure it's merged into 16.0.0-next.3?
> I can't find it in 16.0.0-next.3. When I use @input
> <https://github.com/input>({required: true}) in a directive or a
> component:
>
> error TS2345: Argument of type '{ required: boolean; }' is not assignable to parameter of type 'string'.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/49304#issuecomment-1473678917>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFKBQQNMNGZJ4BVTGIFY76TW4RBWNANCNFSM6AAAAAAVOJMFJU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

e-oz commented:
thanks @kbrilla
